text/0000-box-aliasing.md

```diff
@@ -0,0 +1,96 @@
+- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Explicitly permit unsafe aliasing with `Box<T>`.
+
+# Motivation
+[motivation]: #motivation
+
+The aliasing rules for `Box` are not entirely clear.
+Currently, rustc emits `noalias` on `Box` parameters, which implies some kind of aliasing requirement.
+Attempts to formalize this generate examples which authors of unsafe code who are not intimately familiar with compiler internals find deeply disturbing.
+For example, formalizing this rule in Stacked Borrows says that this program is UB:
+```rust
+let mut b = Box::new(0usize);
+let ptr = &*b as *const usize;
+let a = b;
+unsafe {
+    dbg!(*ptr);
+}
+```
+To emphasize this point: The way users react to learning about this is different from learning about aliasing implications of `&mut`.
+When confronted with this situation, users routinely have all their intuition about `Box<T>` invalidated, as opposed to being led to a greater understanding about the complexity of aliasing optimizations.
+For `&mut`, users are sometimes frustrated by what can be derived from axioms they are familiar with.
+But for `Box<T>`, such users are shocked by the axioms.
+
+In safe Rust, our users cannot avoid learning the rules because the compiler issues hard errors.
+But in the current state of unsafe Rust, we do not even have a single place that anyone can refer users to so that they can learn all the rules.
+And on top of that we need to contend with the reality that while we are still figuring out what exactly the rules for unsafe Rust are, we are (at time of writing) 8 years past 1.0.
+We have 8 years of legacy code and some level of stability obligation to all of that code.
+Optimizing that code based on undocumented and unmentioned aliasing properties is likely to cause problems.
+
+At the time of writing, we are aware of 149 crates published on crates.io which encounter UB due to the aliasing rules for `Box<T>`.
+(Should we post this list somewhere?)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There is nothing to teach here.
+The existing documentation for the type `Box` is sufficient, which is the point of this RFC.
+Some documentation in `std::boxed` which was previously added in an effort to ameliorate the confusion discussed in this PR can now be elided.
+
+# Reference-level explanation
```

Comment:
From Connor on Discord:
Perhaps something like
> This RFC includes a promise that, without otherwise establishing specifics of the aliasing model, it is permitted always to aliasing a `Box<T>` and a raw pointer to the interior of the box that, and that (provided the pointer was not derived from a reference) the pointer may be used until the `Box` is dropped (or the storage is otherwise deallocated), a conflicting reference is created (which at the very least is a mutable reference), or the `Box` is dereferenced and used to write to the storage.